### ISSUE TITLE
Mutli-dimensional timeseries and decode support for RNN autoencoder

### ISSUE THREAD

**Issue by George3d6** [mindsdb/lightwood#169]
`Warning: This issue is intended as a test to be given to candidates for a job at Mindsdb, please do not start working on it, unless we've message you telling you to do so.`

## Current state

We presently have an RNN-based encoder for timeseries data, which will take a sequence of arbitrary length, e.g. `[1, 2, 3, 4, 5, 6, .... , 2000]` and turn it into an embedding of a fixed size (I believe it currently defaults to 4).'

It can be found here:

https://github.com/mindsdb/lightwood/blob/stable/lightwood/encoders/time_series/rnn.py

And the actual network is defined here:

https://github.com/mindsdb/lightwood/blob/stable/lightwood/encoders/time_series/helpers/rnn_helpers.py
 
(Note, there is an `EncoderRNNNumerical` and a `DecoderRNNNumerical`, but only the former is currently in use)

The unit tests for it can be found here:

https://github.com/mindsdb/lightwood/blob/stable/lightwood/encoders/time_series/rnn.py#L154

They can be executed by cloning lightwood, installing the requirements (`pip3 install -r requirements.txt`), adding lightwood to your python path (e.g. `export PYTHONPATH=$PYTHONPATH:/home/<your user>/lightwood`) and then cd-ing into the cloned directory and running `lightwood/encoders/time_series/rnn.py`.

## The task

You need to add a working `decode` function to the encoder, as well as extend the `encode` function to work with multi-dimensional timeseries data. You also need to write appropriate unit tests to validate all this functionality.

### Decoding

The interface for decode should be: `def decode(self, encoded_data)`.

Example usage & testing:
```
data = [[1,2,3,4,5,6],[2,4,8,16,32]]

encoder = RnnEncoder()
encoder.prepare_encoder(self, priming_data=data)
encoded_data = encoder.encode(data)
decoded_data = encoder.decode(encoded_data)

for i in range(len(data)):
    for j in range(len(data[i])):
        assert(decoded_data[i][j] == data[i][j])
```

### Encode Multi-Dimensional time series

A multidimensional time series cell will look something like this: `[(x11,x12... x1n), (x21, x22.... x2n), (xm1, xm2... xmn)]`.

So, for example, assume we have a time samples:

t1, t2, t3,... .tn at uneven spacing form each other (e.g. t1 might be 1000, t2 3000, t3 3200)

We could have the multidimensional time-series:

`[(t2 - t1), sin(t2)), (t3 - t2), sin(t3), .... (tn - tn-1, sin(tn))]`

This should be encoded into `encoded_vector_size` nr of dimensions (by default 4) much like the previous time series, we can check the validity of the embedding by seeing how well the previously implemented `decode` function is able to decode it.

The same `encode` method we have right now should be made to handle this case, without modifying its interface (i.e. adding or removing args from it)

## Bonus objectives

The encoder can train on batches

Assuming that each time series has the same length or, even better, when assuming that each time series can have a variable length.

## Notes

* The specific unit tests to be written are left as an exercise for the implementer, the scope of the unit tests should be to asses performance as well as to see if the autoencoder is doing roughly what would be expected of it. You needn't validate anything about how the resulting embeddings behave when feed further down the line into the mixer.

* Please ask any questions you might have here, unless they need to be private, since that way they can remain here as an FAQ

* Message us when you are done with this project, don't make a PR (since it would be visible to other participants that might not have finished yet)

* Ideally your encoder should run on both GPU and CPU, but if you only had time to make it work on one of them, tell us which

* Feel free to rewrite as much of the existing code as you wish, as long as the underlying functionality stays the same whatever changes you make should be fine.

**Comment by paxcema** [comment]
Hi, I've got some questions:

1. Considering the last note, I suspect there is no problem in using (and augmenting) the ```DecoderRNNNumerical``` class for internal use in the ```RnnEncoder``` at the decoding step, am I right?

2. Is any single multi-dimensional encoder expected to handle arbitrarily dimensioned time series at inference? Or can we assume that each trained instance will expect a fixed number of dimensions at both training and testing time?

3. Once trained, is it valid to condition a query decoding by feeding an initial value in addition to the embedding / hidden state?

**Comment by George3d6** [comment]
1. I'm almost sure the answer is yes since it's only used in the RnnEncoder and this is meant to replace said encoder.

2. Sorry, should have been more specific there (maybe?) the nr of dimensions is constant, the size of the dimensions might not be. So you can expect the vector to always have `n` dimensions but the size might be (10,100,2) or (20,200,4)... though tbh it might be that not all sizes need to vary, only 1 of them. But I'm unsure right now since I wrote this like 1 month ago and have since forgotten about the specific use-cases @torrmal can answer this better than me.

3. Can you expand on what you mean by "an initial value"?

Also, please note the note at the beginning:

```
Warning: This issue is intended as a test to be given to candidates for a job at Mindsdb, please do not start working on it, unless we've message you telling you to do so.
``` 

I'm not sure if @torrmal messaged you to work on this since I don't handle recruiting for this role, but if you want to put in the time to solve this (since it might end up being sort of complicated) it might be worth chatting with him first (unless he's already messaged you about working on this in which case please carry on).

**Comment by paxcema** [comment]
Thanks for the answers @George3d6. Indeed, I've already talked with @torrmal so it's ok for me to try my hand at this.

2. Okay, let's see if @torrmal can help us here. I was wondering about this because if the number of dimensions were not fixed, then the decoder additionally has to determine each time series dimensionality from its embedding.

3. As for the "initial value", I am talking about the input feature tensor that goes into the GRU cell decoder along with the hidden state. To put it another way, imagine the time series I need to decode from its embedding is [2, 4, 8]. What I am wondering is whether it's valid to pass [2] as the initial input tensor to the decoder, along with the embedding. Given the example usage I suspect this is not the case, though.

I have another question:

4. Are there any requirements with respect to the decoder performance for this challenge? e.g. how close should an acceptable reconstruction be, number of parameters, convergence speed, etc.

**Comment by George3d6** [comment]
2. Realistically, what you want to do here, is determine the "max" size you want to work with based on the dataset and pad the vectors with some null token, I think it's fair to assume that no input will be of a size greater than those feed while preparing the encoder (see `prepare_encoder` method and `priming_data` argument).

3. I mean, if the encoder you create generates an embedding that contains said initial value that seems like fair game to me...

4. To be honest, I'm not sure I can say much here, performance is a bonus but we don't have a specific performance benchmark, more like: It should be optimized for performance via using the kind of best practices that are known to increase performance (small size, short backprop time based on the loss+operations, allow for efficient GPU usage including stuff like batching, don't rely on being able to converge with lr=1e-5... etc)

For an example dataset that this should ideally encode in e.g. < 20 minutes on your machine (based on the specs in another issue) see: https://github.com/mindsdb/mindsdb-examples/tree/master/others/twitter

(You have to run the preprocess.py script), but to be honest, I wouldn't start with benchmarks since those aren't the focus, as long as in principle the approach is flexible and allows for the kind of tweaking that can be used to improve performance.

At least that's my opinion, @torrmal *might* have a slightly different take here.
